File: src/live2d_llm_expression/mapper/expression_layer.py
from __future__ import annotations

from live2d_llm_expression.emotion.schema import EmotionIntent, SpecialExpressionName

EXPRESSION_LAYER_KEYS = [
    "ParamExpression_1",
    "ParamExpression_2",
    "ParamExpression_3",
    "ParamExpression_4",
    "ParamHide_EyesL1",
    "ParamHighLightHide_EyesL1",
    "ParamHide_EyeSocket",
    "ParamHide_EyeSocket2",
]

EXPRESSION_LAYER_PRESETS: dict[str, dict[str, float]] = {
    "none": {
        "ParamExpression_1": 0.0,
        "ParamExpression_2": 0.0,
        "ParamExpression_3": 0.0,
        "ParamExpression_4": 0.0,
        "ParamHide_EyesL1": 0.0,
        "ParamHighLightHide_EyesL1": 0.0,
        "ParamHide_EyeSocket": 0.0,
        "ParamHide_EyeSocket2": 0.0,
    },
    "tears": {
        "ParamExpression_1": 1.0,
        "ParamExpression_2": 0.0,
        "ParamExpression_3": 0.0,
        "ParamExpression_4": 0.0,
        "ParamHide_EyesL1": 0.0,
        "ParamHighLightHide_EyesL1": 0.0,
        "ParamHide_EyeSocket": 0.0,
        "ParamHide_EyeSocket2": 0.0,
    },
    "tear_drop": {
        "ParamExpression_1": 0.0,
        "ParamExpression_2": 1.0,
        "ParamExpression_3": 0.0,
        "ParamExpression_4": 0.0,
        "ParamHide_EyesL1": 1.0,
        "ParamHighLightHide_EyesL1": 1.0,
        "ParamHide_EyeSocket": 1.0,
        "ParamHide_EyeSocket2": 1.0,
    },
    "closed_eye_smile": {
        "ParamExpression_1": 0.0,
        "ParamExpression_2": 0.0,
        "ParamExpression_3": 1.0,
        "ParamExpression_4": 0.0,
        "ParamHide_EyesL1": 1.0,
        "ParamHighLightHide_EyesL1": 1.0,
        "ParamHide_EyeSocket": 1.0,
        "ParamHide_EyeSocket2": 1.0,
    },
    "squeezed_eyes": {
        "ParamExpression_1": 0.0,
        "ParamExpression_2": 0.0,
        "ParamExpression_3": 0.0,
        "ParamExpression_4": 1.0,
        "ParamHide_EyesL1": 1.0,
        "ParamHighLightHide_EyesL1": 1.0,
        "ParamHide_EyeSocket": 1.0,
        "ParamHide_EyeSocket2": 1.0,
    },
}


def resolve_special_expression(intent: EmotionIntent) -> SpecialExpressionName:
    if intent.special_expression and intent.special_expression != "none":
        return intent.special_expression
    if intent.emotion == "crying" and intent.intensity >= 0.5:
        return "tears"
    if (
        intent.emotion in {"happy", "teasing"}
        and intent.intensity >= 0.8
        and intent.eyes == "closed_smile"
    ):
        return "closed_eye_smile"
    if intent.emotion == "panic" and intent.intensity >= 0.7:
        return "squeezed_eyes"
    return "none"


def apply_special_expression(
    params: dict[str, float],
    special_expression: str,
) -> dict[str, float]:
    next_params = dict(params)
    next_params.update(EXPRESSION_LAYER_PRESETS.get(special_expression, EXPRESSION_LAYER_PRESETS["none"]))
    return next_params

File: src-ts/expression-layer.ts
import type { NormalizedEmotionIntent, SpecialExpressionName } from "./types.js";

export const EXPRESSION_LAYER_PRESETS: Record<string, Record<string, number>> = {
  none: {
    ParamExpression_1: 0,
    ParamExpression_2: 0,
    ParamExpression_3: 0,
    ParamExpression_4: 0,
    ParamHide_EyesL1: 0,
    ParamHighLightHide_EyesL1: 0,
    ParamHide_EyeSocket: 0,
    ParamHide_EyeSocket2: 0,
  },
  tears: {
    ParamExpression_1: 1,
    ParamExpression_2: 0,
    ParamExpression_3: 0,
    ParamExpression_4: 0,
    ParamHide_EyesL1: 0,
    ParamHighLightHide_EyesL1: 0,
    ParamHide_EyeSocket: 0,
    ParamHide_EyeSocket2: 0,
    ParamCryDown_L: 0.32,
    ParamTearDown_1: 0.55,
    ParamTearDown_2: 0.36,
    ParamTearDown_3: 0.2,
  },
  tear_drop: {
    ParamExpression_1: 0,
    ParamExpression_2: 1,
    ParamExpression_3: 0,
    ParamExpression_4: 0,
    ParamHide_EyesL1: 1,
    ParamHighLightHide_EyesL1: 1,
    ParamHide_EyeSocket: 1,
    ParamHide_EyeSocket2: 1,
    ParamTearDown_1: 0.3,
  },
  closed_eye_smile: {
    ParamExpression_1: 0,
    ParamExpression_2: 0,
    ParamExpression_3: 1,
    ParamExpression_4: 0,
    ParamHide_EyesL1: 1,
    ParamHighLightHide_EyesL1: 1,
    ParamHide_EyeSocket: 1,
    ParamHide_EyeSocket2: 1,
    ParamEyeCircles: 0,
  },
  squeezed_eyes: {
    ParamExpression_1: 0,
    ParamExpression_2: 0,
    ParamExpression_3: 0,
    ParamExpression_4: 1,
    ParamHide_EyesL1: 1,
    ParamHighLightHide_EyesL1: 1,
    ParamHide_EyeSocket: 1,
    ParamHide_EyeSocket2: 1,
    ParamEyeCircles: 0.18,
  },
};

export function resolveSpecialExpression(intent: NormalizedEmotionIntent): SpecialExpressionName {
  if (intent.specialExpression && intent.specialExpression !== "none") {
    return intent.specialExpression;
  }
  if (intent.emotion === "crying" && intent.intensity >= 0.5) return "tears";
  if (
    (intent.emotion === "happy" || intent.emotion === "teasing") &&
    intent.intensity >= 0.8 &&
    (
      intent.eyes === "closed_smile"
      || intent.tone === "proud"
      || intent.tone === "relieved"
      || intent.tone === "grateful"
    )
  ) {
    return "closed_eye_smile";
  }
  return "none";
}

export function applySpecialExpression(
  params: Record<string, number>,
  specialExpression: string,
): Record<string, number> {
  return {
    ...params,
    ...(EXPRESSION_LAYER_PRESETS[specialExpression] ?? EXPRESSION_LAYER_PRESETS.none),
  };
}
